openpgpjs/openpgpjs
# Basic Deno support

Related to Discussion #1447

Changes made:
  * Update README.md to indicate experimental support for Deno
  * Add extra checks in ```src/*```, to avoid directly accessing objects/properties not defined in a Deno environment, specifically ```navigator.userAgent``` ~~and ```navigator.hardwareCurrency```~~

Tested with:
*   ```npm test```
*   ```npm run browsertest```
*   ```npm run lint```
*   Example 1-3 from [README.md#examples](https://github.com/openpgpjs/openpgpjs/#examples), using Deno 1.16.4, both JavaScript and TypeScript mode

Testable by:

1.  Make sure openpgpjs is checked out locally
2.  If not already done - run ```npm run build``` to generate ```dist/openpgp.mjs```
2.  Create a test application - ```test.js``` - (in this example in a directory adjacent to openpgpjs ```../denotest```)
```js
// Import locally
import * as openpgp from '../openpgpjs/dist/openpgp.mjs';

(async () => {
    const message = await openpgp.createMessage({ binary: new Uint8Array([0x01, 0x01, 0x01]) });
    const encrypted = await openpgp.encrypt({
        message, // input as Message object
        passwords: ['secret stuff'], // multiple passwords possible
        format: 'binary' // don't ASCII armor (for Uint8Array output)
    });
    console.log(encrypted); // Uint8Array

    const encryptedMessage = await openpgp.readMessage({
        binaryMessage: encrypted // parse encrypted bytes
    });
    const { data: decrypted } = await openpgp.decrypt({
        message: encryptedMessage,
        passwords: ['secret stuff'], // decrypt with password
        format: 'binary' // output as Uint8Array
    });
    console.log(decrypted); // Uint8Array([0x01, 0x01, 0x01])
})();
```
3. Run test application ```deno run test.js```
4. It works (or should, at least)!
5. If implemented, this would be a fully working deno application
```js
import * as openpgp from '<deno land url>/openpgp/dist/openpgp.mjs';
p
(async () => {
    const message = await openpgp.createMessage({ binary: new Uint8Array([0x01, 0x01, 0x01]) });
    const encrypted = await openpgp.encrypt({
        message, // input as Message object
        passwords: ['secret stuff'], // multiple passwords possible
        format: 'binary' // don't ASCII armor (for Uint8Array output)
    });
    console.log(encrypted); // Uint8Array

    const encryptedMessage = await openpgp.readMessage({
        binaryMessage: encrypted // parse encrypted bytes
    });
    const { data: decrypted } = await openpgp.decrypt({
        message: encryptedMessage,
        passwords: ['secret stuff'], // decrypt with password
        format: 'binary' // output as Uint8Array
    });
    console.log(decrypted); // Uint8Array([0x01, 0x01, 0x01])
})();
```
